### ISSUE TITLE
how to make traindata?

### ISSUE THREAD

**Issue by wuxiaolianggit** [zhangzjn/APB2Face#5]
excuseme,how to make traindata? @zhangzjn

**Comment by zhangzjn** [comment]
You can refer Chapter 3 of the paper and closed issues for details, and debug t7 file to find the data structure. You might use some libs, e.g. 'python_speech_feature', 'librosa', and 'scipy.io.wav'.

Some code for the reference:
```
def feature_extraction_AnnVI(video_file, audio_file, label_file, out_path, img_size=256):
    # audio setting
    n_fft = 2048  # 44100/30 1470
    hop_length = 512  # 44100/60 735
    n_mfcc = 20
    sr = 44100
    win_size = 64
    cap = cv2.VideoCapture(video_file)
    fps = cap.get(cv2.CAP_PROP_FPS)
    cap.release()

    # load / solve data
    sig, rate = librosa.load(audio_file, sr=sr, duration=None)
    time_duration = len(sig) / rate
    # print('time of duration : {}'.format(time_duration))l
    f_mfcc = librosa.feature.mfcc(sig, rate, n_mfcc=n_mfcc, n_fft=n_fft, hop_length=hop_length)
    f_mfcc_delta = psf.base.delta(f_mfcc, 2)
    f_mfcc_delta2 = psf.base.delta(f_mfcc_delta, 2)
    f_mfcc_all = np.concatenate((f_mfcc, f_mfcc_delta, f_mfcc_delta2), axis=0)
```
I will release the related code in the next version.

**Comment by wuxiaolianggit** [comment]
Thank you for your reply

**Comment by QUTGXX** [comment]
> You can refer Chapter 3 of the paper and closed issues for details, and debug t7 file to find the data structure. You might use some libs, e.g. 'python_speech_feature', 'librosa', and 'scipy.io.wav'.
> 
> Some code for the reference:
> 
> ```
> def feature_extraction_AnnVI(video_file, audio_file, label_file, out_path, img_size=256):
>     # audio setting
>     n_fft = 2048  # 44100/30 1470
>     hop_length = 512  # 44100/60 735
>     n_mfcc = 20
>     sr = 44100
>     win_size = 64
>     cap = cv2.VideoCapture(video_file)
>     fps = cap.get(cv2.CAP_PROP_FPS)
>     cap.release()
> 
>     # load / solve data
>     sig, rate = librosa.load(audio_file, sr=sr, duration=None)
>     time_duration = len(sig) / rate
>     # print('time of duration : {}'.format(time_duration))l
>     f_mfcc = librosa.feature.mfcc(sig, rate, n_mfcc=n_mfcc, n_fft=n_fft, hop_length=hop_length)
>     f_mfcc_delta = psf.base.delta(f_mfcc, 2)
>     f_mfcc_delta2 = psf.base.delta(f_mfcc_delta, 2)
>     f_mfcc_all = np.concatenate((f_mfcc, f_mfcc_delta, f_mfcc_delta2), axis=0)
> ```
> 
> I will release the related code in the next version.

Hi, I am interested in your great work. I want to know if we get the landmark mask based on face++, should we train the first stage network (audio2landmark)?

**Comment by zhangzjn** [comment]
font{
        line-height: 1.6;
    }
    ul,ol{
        padding-left: 20px;
        list-style-position: inside;
    }


    
    
        You can refer the repo "https://github.com/FacePlusPlus/facepp-python-sdk" to get the landmark for your own datasets, and you have to firstly train the `audio2landmark` module.Good luck.
    
        
            
        
    
    
     
    
        
                    
            
                
                    
                            
                                
                            
                            
                                vtzhang
                                    Ph.D.
                            
                            
                                
                            
                    
                        
                            
                                    Zhejiang University, APRIL
                                    +86 17342017381
                                    zhangjiangning@zju.edu.cn
                                    38 Zheda Rd, Hangzhou 310058, P.R. China
                                    310000
 

    On 8/14/2020 10:49，XINXIN GE<notifications@github.com> wrote： 




You can refer Chapter 3 of the paper and closed issues for details, and debug t7 file to find the data structure. You might use some libs, e.g. 'python_speech_feature', 'librosa', and 'scipy.io.wav'.
Some code for the reference:
def feature_extraction_AnnVI(video_file, audio_file, label_file, out_path, img_size=256):
    # audio setting
    n_fft = 2048  # 44100/30 1470
    hop_length = 512  # 44100/60 735
    n_mfcc = 20
    sr = 44100
    win_size = 64
    cap = cv2.VideoCapture(video_file)
    fps = cap.get(cv2.CAP_PROP_FPS)
    cap.release()

    # load / solve data
    sig, rate = librosa.load(audio_file, sr=sr, duration=None)
    time_duration = len(sig) / rate
    # print('time of duration : {}'.format(time_duration))l
    f_mfcc = librosa.feature.mfcc(sig, rate, n_mfcc=n_mfcc, n_fft=n_fft, hop_length=hop_length)
    f_mfcc_delta = psf.base.delta(f_mfcc, 2)
    f_mfcc_delta2 = psf.base.delta(f_mfcc_delta, 2)
    f_mfcc_all = np.concatenate((f_mfcc, f_mfcc_delta, f_mfcc_delta2), axis=0)

I will release the related code in the next version.

Hi, I am interested in your great work. I want to know if we get the landmark mask based on face++, should we train the first stage network (audio2landmark)?

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.

**Comment by wuxiaolianggit** [comment]
您好，大神，想问你一个问题，我利用你提供的方法去制作数据集，制作t7数据集的时候，landmark的顺序对训练结果有没有影响，还是只要求出106个点就可以了，不用在意这些点的顺序？ @zhangzjn

**Comment by zhangzjn** [comment]
font{
        line-height: 1.6;
    }
    ul,ol{
        padding-left: 20px;
        list-style-position: inside;
    }


    
    landmark送入网络时要保持一个固定的顺序
        
    
        
            
        
    
    
     
    
        
                    
            
                
                    
                            
                                
                            
                            
                                vtzhang
                                    Ph.D.
                            
                            
                                
                            
                    
                        
                            
                                    Zhejiang University, APRIL
                                    +86 17342017381
                                    zhangjiangning@zju.edu.cn
                                    38 Zheda Rd, Hangzhou 310058, P.R. China
                                    310000
 

    On 8/14/2020 15:28，wuxiaolianggit<notifications@github.com> wrote： 




您好，大神，想问你一个问题，我利用你提供的方法去制作数据集，制作t7数据集的时候，landmark的顺序对训练结果有没有影响，还是只要求出106个点就可以了，不用在意这些点的顺序？ @zhangzjn



—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.

**Comment by wuxiaolianggit** [comment]
就是说保持一个固定的顺序就可以了，并不需要和您的数据顺序一样，是这样吗？大神。 @zhangzjn

**Comment by zhangzjn** [comment]
font{
        line-height: 1.6;
    }
    ul,ol{
        padding-left: 20px;
        list-style-position: inside;
    }


    
    
        是的，可以按照你自己定义的数据顺序，当然你也可以使用68个点或者更多点的landmark作为面部表情的控制信息
        
    
        
            
        
    
    
     
    
        
                    
            
                
                    
                            
                                
                            
                            
                                vtzhang
                                    Ph.D.
                            
                            
                                
                            
                    
                        
                            
                                    Zhejiang University, APRIL
                                    +86 17342017381
                                    zhangjiangning@zju.edu.cn
                                    38 Zheda Rd, Hangzhou 310058, P.R. China
                                    310000
 

    On 8/14/2020 16:08，wuxiaolianggit<notifications@github.com> wrote： 




就是说保持一个固定的顺序就可以了，并不需要和您的数据顺序一样，是这样吗？大神。 @zhangzjn



—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.

**Comment by wuxiaolianggit** [comment]
好的，了解了，多谢大神的解答啊 @zhangzjn

**Comment by ZerRui** [comment]
import cv2
import librosa
import numpy as np
import torch
import python_speech_features as psf

def interpolate_features(features, input_rate, output_rate, output_len=None):
    num_features = features.shape[1]
    input_len = features.shape[0]
    seq_len = input_len / float(input_rate)
    if output_len is None:
        output_len = int(seq_len * output_rate)
    input_timestamps = np.arange(input_len) / float(input_rate)
    output_timestamps = np.arange(output_len) / float(output_rate)
    output_features = np.zeros((output_len, num_features))
    for feat in range(num_features):
        output_features[:, feat] = np.interp(output_timestamps,
                                             input_timestamps,
                                             features[:, feat])
    return output_features

def add_chunk_feature(resample_features):
    audio_window_size = 64
    audio_window_stride = 1
    zero_pad = np.zeros((int(audio_window_size / 2), resample_features.shape[1]))
    network_output = np.concatenate((zero_pad, resample_features, zero_pad), axis=0)
    windows = []
    for window_index in range(0, network_output.shape[0] - audio_window_size, audio_window_stride):
        windows.append(network_output[window_index:window_index + audio_window_size])

    return windows


def feature_extraction_AnnVI(audio_file):
    # audio setting
    n_fft = 2048  # 44100/30 1470
    hop_length = 512  # 44100/60 735
    n_mfcc = 20
    sr = 44100
    win_size = 64
    # cap = cv2.VideoCapture(video_file)
    # fps = cap.get(cv2.CAP_PROP_FPS)
    # cap.release()

    # load / solve data
    sig, rate = librosa.load(audio_file, sr=sr, duration=None)
    time_duration = len(sig) / rate
    # print('time of duration : {}'.format(time_duration))l
    f_mfcc = librosa.feature.mfcc(sig, rate, n_mfcc=n_mfcc, n_fft=n_fft, hop_length=hop_length)
    f_mfcc_delta = psf.base.delta(f_mfcc, 2)
    f_mfcc_delta2 = psf.base.delta(f_mfcc_delta, 2)
    f_mfcc_all = np.concatenate((f_mfcc, f_mfcc_delta, f_mfcc_delta2), axis=0)
    f_mfcc_all_T = np.transpose(f_mfcc_all)
    num_frames = int(round(len(sig)/rate*30))
    resample_features = interpolate_features(f_mfcc_all_T,50,30,output_len=num_frames)
    finally_feature = add_chunk_feature(resample_features)
    return finally_feature

if __name__ == '__main__':
    wav_path = '/home/zml/obama01.m4a'
    finally_feature = feature_extraction_AnnVI(wav_path)
    print(len(finally_fature))

大神，mfcc特征求完了之后，我用这种方法进行重采样和加窗处理可以吗？目前用5000左右的图片，训练效果不好，您可以帮忙看下音频特征这边是不是有问题啊？

**Comment by ZerRui** [comment]
`import cv2
import librosa
import numpy as np
import torch
import python_speech_features as psf

def interpolate_features(features, input_rate, output_rate, output_len=None):
    num_features = features.shape[1]
    input_len = features.shape[0]
    seq_len = input_len / float(input_rate)
    if output_len is None:
        output_len = int(seq_len * output_rate)
    input_timestamps = np.arange(input_len) / float(input_rate)
    output_timestamps = np.arange(output_len) / float(output_rate)
    output_features = np.zeros((output_len, num_features))
    for feat in range(num_features):
        output_features[:, feat] = np.interp(output_timestamps,
                                             input_timestamps,
                                             features[:, feat])
    return output_features

def add_chunk_feature(resample_features):
    audio_window_size = 64
    audio_window_stride = 1
    zero_pad = np.zeros((int(audio_window_size / 2), resample_features.shape[1]))
    network_output = np.concatenate((zero_pad, resample_features, zero_pad), axis=0)
    windows = []
    for window_index in range(0, network_output.shape[0] - audio_window_size, audio_window_stride):
        windows.append(network_output[window_index:window_index + audio_window_size])

    return windows


def feature_extraction_AnnVI(audio_file):
    # audio setting
    n_fft = 2048  # 44100/30 1470
    hop_length = 512  # 44100/60 735
    n_mfcc = 20
    sr = 44100
    win_size = 64
    # cap = cv2.VideoCapture(video_file)
    # fps = cap.get(cv2.CAP_PROP_FPS)
    # cap.release()

    # load / solve data
    sig, rate = librosa.load(audio_file, sr=sr, duration=None)
    time_duration = len(sig) / rate
    # print('time of duration : {}'.format(time_duration))l
    f_mfcc = librosa.feature.mfcc(sig, rate, n_mfcc=n_mfcc, n_fft=n_fft, hop_length=hop_length)
    f_mfcc_delta = psf.base.delta(f_mfcc, 2)
    f_mfcc_delta2 = psf.base.delta(f_mfcc_delta, 2)
    f_mfcc_all = np.concatenate((f_mfcc, f_mfcc_delta, f_mfcc_delta2), axis=0)
    f_mfcc_all_T = np.transpose(f_mfcc_all)
    num_frames = int(round(len(sig)/rate*30))
    resample_features = interpolate_features(f_mfcc_all_T,50,30,output_len=num_frames)
    finally_feature = add_chunk_feature(resample_features)
    return finally_feature
def load_t7(t7_path):
    a = torch.load(t7_path)
    for k, v in a.items():
        # print(k, len(v))
        # print(k, np.array(v[1]).shape, v[0])
        if k=='audio_features':
            print(v[0])
if __name__ == '__main__':
    wav_path = '/home/zml/obama01.m4a'
    finally_feature = feature_extraction_AnnVI(wav_path)
    print(len(finally_fature))`

**Comment by zhangzjn** [comment]
@ZerRui 您好，已经微信回复了
